share/ipld/namespace_data.go

```diff
@@ -0,0 +1,155 @@
+package ipld
+
+import (
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+)
+
+// Option is the functional option that is applied to the NamespaceData instance
+// to configure data that needs to be stored.
+type Option func(*NamespaceData)
+
+// WithLeaves option specifies that leaves should be collected during retrieval.
+func WithLeaves() Option {
+	return func(data *NamespaceData) {
+		// we over-allocate space for leaves since we do not know how many we will find
+		// on the level above, the length of the Row is passed in as maxShares
+		data.leaves = make([]ipld.Node, data.maxShares)
+	}
+}
+
+// WithProofs option specifies that proofs should be collected during retrieval.
+func WithProofs() Option {
+	return func(data *NamespaceData) {
+		data.proofContainer = newProofCollector(data.maxShares)
+	}
+}
+
+// NamespaceData saves all leaves under the given namespace in the given namespace with corresponded proofs.
+type NamespaceData struct {
+	leaves         []ipld.Node
+	proofContainer *proofCollector
```

Comment:
I know this was the old name, but I just had a thought: Every field is a "container" - maybe we should just name the field to "proofs"